Sources/Crust/Parser.swift

```diff
@@ -42,4 +50,543 @@ public class Parser {
+      pieces.append(piece)
+    }
+    return TypedParameterListSyntax(elements: pieces)
+  }
+
+  func parseTypedParameter() throws -> TypedParameterSyntax {
+    let leftParen = try consume(.leftParen)
+    let ascription = try parseAscription()
+    let rightParen = try consume(.rightParen)
+    return TypedParameterSyntax(
+      leftParenToken: leftParen,
+      ascription: ascription,
+      rightParenToken: rightParen
+    )
+  }
+
+  func parseTypeIndices() throws -> TypeIndicesSyntax {
+    let colon = try consume(.colon)
+    let expr = try parseExpr()
+    return TypeIndicesSyntax(colonToken: colon, indexExpr: expr)
+  }
+
+  func parseAscription() throws -> AscriptionSyntax {
+    let boundNames = try parseIdentifierList()
+    let colonToken = try consume(.colon)
+    let expr = try parseExpr()
+    return AscriptionSyntax(
+      boundNames: boundNames,
+      colonToken: colonToken,
+      typeExpr: expr
+    )
+  }
+}
+
+extension Parser {
+  func parseDataDeclaration() throws -> DataDeclSyntax {
+    do {
+      let dataTok = try consume(.dataKeyword)
+      let dataId = try parseQualifiedNamePiece()
+      let paramList = parseTypedParameterList()
+      let indices = try parseTypeIndices()
+      let whereTok = try consume(.whereKeyword)
+      let constrList = parseConstructorList()
+      return DataDeclSyntax(
+        dataToken: dataTok,
+        dataIdentifier: dataId.name,
+        typedParameterList: paramList,
+        typeIndices: indices,
+        whereToken: whereTok,
+        constructorList: constrList
+      )
+    } catch let e {
+      throw e
+    }
+  }
+
+  func parseConstructorList() -> ConstructorListSyntax {
+    var pieces = [ConstructorDeclSyntax]()
+    while let piece = try? parseConstructor() {
+      pieces.append(piece)
+    }
+    return ConstructorListSyntax(elements: pieces)
+  }
+
+  func parseConstructor() throws -> ConstructorDeclSyntax {
+    let pipe = try consume(.pipe)
+    let ascription = try parseAscription()
+    return ConstructorDeclSyntax(pipeToken: pipe, ascription: ascription)
+  }
+}
+
+extension Parser {
+  func parseFunctionDecl() throws -> FunctionDeclSyntax {
+    let ascription = try parseAscription()
+    let ascSemicolon = try consume(.semicolon)
+    let clauses = try parseFunctionClauseList()
+    let clausesSemiColon = try consume(.semicolon)
+    return FunctionDeclSyntax(
+      ascription: ascription,
+      ascriptionSemicolon: ascSemicolon,
+      clauseList: clauses,
+      trailingSemicolon: clausesSemiColon
+    )
+  }
+
+  func parseFunctionClauseList() throws -> FunctionClauseListSyntax {
+    var pieces = [FunctionClauseSyntax]()
+    while let piece = try? parseFunctionClause() {
+      pieces.append(piece)
+    }
+
+    guard !pieces.isEmpty else {
+      throw ParseError.failedParsingFunctionClause
+    }
+    return FunctionClauseListSyntax(elements: pieces)
+  }
+
+  func parseFunctionClause() throws -> FunctionClauseSyntax {
+    return try split({
```

Comment:
This pattern is also significantly more difficult to debug.

Reply:
We can't know whether or not we have a `with` clause until we've parsed through the entire LHS.  I can collapse `WithRuleFunctionClause` and `NormalFunctionClause` and just make the with clause optional.

Reply:
Never mind, can't do that.  We don't define `SyntaxKind`s for abstract derived classes of `Syntax` and so you can't have an `Expr?` node

Reply:
This is fine, then. 👍